### PR TITLE
refactor(image_decoder): refactor image decoder and image cache

### PR DESCRIPTION
**Pull request by W-Mai**
### Description of the feature or fix

Move image cache callbacks to image cache module.

Because there were some dependencies before, there was no way to put these callbacks into the image cache module. Now it can do this. This makes the image decoder module code clearer and more concise.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.10](https://github.com/szepeviktor/astyle/releases/tag/v3.4.10) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: src/misc/cache/lv_image_cache.c
Comment (PGNetHun) [inline src/misc/cache/lv_image_cache.c:1]: I wonder whether separating the image cache and image header cache into 2 files is worth:
- `lv_image_cache.c` / `.h`
- `lv_image_header_cache.c` / `.h`

So if one of these cache implementations will be fixed/refactored, then only that file will be changed.
The other advantage of separating image and header cache implementations into 2 separate files:
you need to add `#if LV_CACHE_DEF_SIZE > 0` ( or `#if LV_IMAGE_HEADER_CACHE_DEF_CNT > 0`) only ONCE in the file beginning, and if cache is disabled for either one, then none of the image/header caching functions will be compiled into the firmware.
So, basically we could get rid of the many `#if LV_CACHE_DEF_SIZE > 0` and `#if LV_IMAGE_HEADER_CACHE_DEF_CNT > 0` lines in the actual implementation, and add them only once in their own source files.

File: src/draw/lv_image_decoder.c [hunk 1]
```diff
@@ -70,24 +61,8 @@ void _lv_image_decoder_init(void)
...
-
-#if LV_IMAGE_HEADER_CACHE_DEF_CNT > 0
-    img_header_cache_p = lv_cache_create(&lv_cache_class_lru_rb_count,
-    sizeof(lv_image_header_cache_data_t), LV_IMAGE_HEADER_CACHE_DEF_CNT, (lv_cache_ops_t) {
-        .compare_cb = (lv_cache_compare_cb_t)image_decoder_header_cache_compare_cb,
-        .create_cb = NULL,
-        .free_cb = (lv_cache_free_cb_t)image_decoder_header_cache_free_cb
-    });
-#endif
+    lv_image_cache_init();
```
Comment (PGNetHun) [inline src/draw/lv_image_decoder.c:66]: I think these 2 inits could be inside `#if LV_CACHE_DEF_SIZE > 0`   and `#if LV_IMAGE_HEADER_CACHE_DEF_CNT > 0`, so init functions are called ONLY when cache is really enabled.

File: src/misc/cache/lv_image_cache.c [hunk 2]
```diff
@@ -37,6 +49,38 @@
...
+        .compare_cb = (lv_cache_compare_cb_t) image_header_cache_compare_cb,
+        .create_cb = NULL,
+        .free_cb = (lv_cache_free_cb_t) image_header_cache_free_cb
+    });
+
+    return img_header_cache_p != NULL;
+#else
+    return false;
+#endif
+}
```
Comment (liamHowatt) [inline src/misc/cache/lv_image_cache.c]: Would it make sense to add `if(img_cache_p != NULL) return true;` and `if(img_header_cache_p != NULL) return true;` at the beginning since these functions are part of a public API? They might be called a second time by something other than the image decoder.

File: src/misc/cache/lv_image_cache.h [hunk 3]
```diff
@@ -27,6 +27,18 @@ extern "C" {
  * GLOBAL PROTOTYPES
  **********************/
 
+/**
+ * Initialize image cache.
+ * @return true: initialization succeeded, false: failed.
+ */
+bool lv_image_cache_init(void);
```
Comment (XuNeo) [inline src/misc/cache/lv_image_cache.h]: please use lv_result_t as return type.

File: src/misc/cache/lv_image_cache.c (hunk 2 above)
Comment (W-Mai) [reply src/misc/cache/lv_image_cache.c]: In fact, it should not be called by other functions, but your suggestion is very good. I'll add it!

Comment (W-Mai) [reply src/misc/cache/lv_image_cache.c]: Done

File: src/misc/cache/lv_image_cache.h (hunk 3 above)
Comment (W-Mai) [reply src/misc/cache/lv_image_cache.h]: Done

**Review by XuNeo — DISMISSED** [review]

**Review by PGNetHun — CHANGES_REQUESTED** [review]

File: src/misc/cache/lv_image_cache.c
Comment (W-Mai) [reply src/misc/cache/lv_image_cache.c:1]: This makes sense. Let me update it.

File: src/draw/lv_image_decoder.c (hunk 1 above)
Comment (W-Mai) [reply src/draw/lv_image_decoder.c:66]: Done!

File: src/misc/cache/lv_image_cache.c
Comment (W-Mai) [reply src/misc/cache/lv_image_cache.c:1]: Done!

File: src/misc/cache/lv_image_cache.c [hunk 4]
```diff
@@ -59,51 +94,50 @@ void lv_image_cache_drop(const void * src)
 #endif
 }
 
-void lv_image_cache_resize(uint32_t new_size, bool evict_now)
-{
+/**********************
+ *   STATIC FUNCTIONS
+ **********************/
+
 #if LV_CACHE_DEF_SIZE > 0
```
Comment (liamHowatt) [inline src/misc/cache/lv_image_cache.c:102]: Should there be just one of these in this file instead of many?

File: src/misc/cache/lv_image_header_cache.c [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    LV_UNUSED(src);
+#endif
+}
+
+/**********************
+ *   STATIC FUNCTIONS
+ **********************/
+
+
+#if LV_IMAGE_HEADER_CACHE_DEF_CNT > 0
```
Comment (liamHowatt) [inline src/misc/cache/lv_image_header_cache.c:99]: Should there be just one of these in this file instead of many?

File: src/misc/cache/lv_image_header_cache.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+lv_result_t lv_image_header_cache_init(void);
+
+/**
+ * Resize image header cache.
+ * @param new_size  new size of the cache in count of image headers.
+ * @param evict_now true: evict the image headers should be removed by the eviction policy, false: wait for the next cache cleanup.
+ */
+void lv_image_header_cache_resize(uint32_t new_size, bool evict_now);
+
+#endif /*LV_IMAGE_HEADER_CACHE_DEF_CNT > 0*/
```
Comment (liamHowatt) [inline src/misc/cache/lv_image_header_cache.h:47]: How come this guard doesn't capture the last prototype in this header?

File: src/misc/cache/lv_image_cache.h [hunk 7]
```diff
@@ -40,19 +42,13 @@ void lv_image_cache_drop(const void * src);
  */
 void lv_image_cache_resize(uint32_t new_size, bool evict_now);
 
-/**
- * Invalidate image header cache. Use NULL to invalidate all image headers.
- * It's also automatically called when an image is invalidated.
- * @param src pointer to an image source.
- */
-void lv_image_header_cache_drop(const void * src);
+#endif /*LV_CACHE_DEF_SIZE > 0*/
```
Comment (liamHowatt) [inline src/misc/cache/lv_image_cache.h:47]: How come this guard doesn't capture the last prototype in this header?

**Review by PGNetHun — DISMISSED** [review]
Great, thanks!
I see you kept the cache drop functions to have the code "backward compatible", thx!

File: src/misc/cache/lv_image_cache.h [hunk 8]
```diff
@@ -15,6 +15,8 @@ extern "C" {
  *********************/
 #include "lv_cache_private.h"
```
Comment (liamHowatt) [inline src/misc/cache/lv_image_cache.h]: Can this be moved to the .c file? https://github.com/lvgl/lvgl/issues/5899#issuecomment-2004582233

File: src/misc/cache/lv_image_header_cache.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#define LV_IMAGE_HEADER_CACHE_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*********************
+ *      INCLUDES
+ *********************/
+#include "lv_cache_private.h"
```
Comment (liamHowatt) [inline src/misc/cache/lv_image_header_cache.h]: Can this be moved to the .c file? https://github.com/lvgl/lvgl/issues/5899#issuecomment-2004582233

**Review by liamHowatt — COMMENTED** [review]
This is mainly a review of the changes in response to @PGNetHun's review.

File: src/misc/cache/lv_image_header_cache.h (hunk 6 above)
Comment (liamHowatt) [reply src/misc/cache/lv_image_header_cache.h:47]: https://github.com/lvgl/lvgl/pull/5890#pullrequestreview-1946077546 Ah, I see, please disregard.

File: src/misc/cache/lv_image_cache.h (hunk 7 above)
Comment (liamHowatt) [reply src/misc/cache/lv_image_cache.h:47]: https://github.com/lvgl/lvgl/pull/5890#pullrequestreview-1946077546 Ah, I see, please disregard.

File: src/misc/cache/lv_image_cache.h (hunk 8 above)
Comment (liamHowatt) [reply src/misc/cache/lv_image_cache.h]: `<stdbool.h>` will be needed here fyi, if you do move it.

File: src/misc/cache/lv_image_header_cache.h (hunk 9 above)
Comment (liamHowatt) [reply src/misc/cache/lv_image_header_cache.h]: `<stdbool.h>` will be needed here fyi, if you do move it.

Comment (W-Mai) [reply src/misc/cache/lv_image_header_cache.h]: The LV_CACHE_DEF macro need be imported from this header

Comment (liamHowatt) [reply src/misc/cache/lv_image_header_cache.h]: Gotcha, thanks!

Comment (XuNeo) [reply src/misc/cache/lv_image_header_cache.h]: > The LV_CACHE_DEF macro need be imported from this header

Do you mean `LV_CACHE_DEF_SIZE`? It should be imported by lv_conf_internal.h. Let's keep the private headers private.

Comment (W-Mai) [reply src/misc/cache/lv_image_header_cache.h]: You're right!
Let me update the codes!

Comment (W-Mai) [reply src/misc/cache/lv_image_header_cache.h]: Done!

@liamHowatt Sure! I've updated the codes now. https://github.com/lvgl/lvgl/pull/5890#discussion_r1530297242

File: src/misc/cache/lv_image_cache.h [hunk 10]
```diff
@@ -13,7 +13,11 @@ extern "C" {
 /*********************
  *      INCLUDES
  *********************/
-#include "lv_cache_private.h"
+
+#include "../../lv_conf_internal.h"
+#include <stdbool.h>
```
Comment (XuNeo) [inline src/misc/cache/lv_image_cache.h:18]: Need to remove it and include lv_types.h where lv_result_t is defined.

File: src/misc/cache/lv_image_header_cache.h [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*********************
+ *      INCLUDES
+ *********************/
+
+#include "../../lv_conf_internal.h"
+#include <stdbool.h>
```
Comment (XuNeo) [inline src/misc/cache/lv_image_header_cache.h:18]: ditto.

File: src/misc/cache/lv_image_cache.h (hunk 10 above)
Comment (W-Mai) [reply src/misc/cache/lv_image_cache.h:18]: lv_types.h doesn't supply `bool`

Comment (XuNeo) [reply src/misc/cache/lv_image_cache.h:18]: Emmm. https://github.com/lvgl/lvgl/pull/5767 is still not merged.

Comment (W-Mai) [reply src/misc/cache/lv_image_cache.h:18]: I think if this pr is not merge, we can update this fix later.

**Review by XuNeo — APPROVED** [review]

**Comment by liamHowatt** [comment]
Just this now https://github.com/lvgl/lvgl/pull/5890#discussion_r1530266969

**Review by PGNetHun — APPROVED** [review]

File: src/misc/cache/lv_image_cache.c (hunk 4 above)
Comment (XuNeo) [reply src/misc/cache/lv_image_cache.c:102]: @liamHowatt Could you please explain the issue in detail? I don't understand your point.
